src/workerd/api/node/buffer.c++

```diff
@@ -0,0 +1,396 @@
+  (((x) & 0x000000FF00000000ull) >> 8)  |                                     \
+  (((x) & 0x00000000FF000000ull) << 8)  |                                     \
+  (((x) & 0x0000000000FF0000ull) << 24) |                                     \
+  (((x) & 0x000000000000FF00ull) << 40) |                                     \
+  (((x) & 0x00000000000000FFull) << 56)
+#endif
+
+namespace workerd::api::node {
+
+namespace {
+template <typename T>
+void SwapBytes(kj::ArrayPtr<kj::byte> bytes) {
+  KJ_DASSERT((bytes.size() % sizeof(T)) == 0);
+  uint32_t len = bytes.size() / sizeof(T);
+  T* data = reinterpret_cast<T*>(bytes.begin());
+  for (uint32_t i = 0; i < len; i++) {
+    if constexpr (kj::isSameType<T, uint16_t>()) {
+      data[i] = BSWAP_2(data[i]);
+    } else if constexpr (kj::isSameType<T, uint32_t>()) {
+      data[i] = BSWAP_4(data[i]);
+    } else {
+      data[i] = BSWAP_8(data[i]);
+    }
+  }
+}
+
+enum class Encoding {
+  ASCII,
+  LATIN1,
+  UTF8,
+  UTF16LE,
+  BASE64,
+  BASE64URL,
+  HEX,
+};
+
+Encoding getEncoding(kj::StringPtr encoding) {
+  if (encoding == "utf8"_kj) {
+    return Encoding::UTF8;
+  } else if (encoding == "ascii") {
+    return Encoding::ASCII;
+  } else if (encoding == "latin1") {
+    return Encoding::LATIN1;
+  } else if (encoding == "utf16le") {
+    return Encoding::UTF16LE;
+  } else if (encoding == "base64") {
+    return Encoding::BASE64;
+  } else if (encoding == "base64url") {
+    return Encoding::BASE64URL;
+  } else if (encoding == "hex") {
+    return Encoding::HEX;
+  }
+
+  KJ_UNREACHABLE;
+}
+
+kj::Maybe<uint> tryFromHexDigit(char c) {
+  if ('0' <= c && c <= '9') {
+    return c - '0';
+  } else if ('a' <= c && c <= 'f') {
+    return c - ('a' - 10);
+  } else if ('A' <= c && c <= 'F') {
+    return c - ('A' - 10);
+  } else {
+    return nullptr;
+  }
+}
+
+kj::Array<byte> decodeHexTruncated(kj::ArrayPtr<const char> text) {
+  // We do not use kj::decodeHex because we need to match Node.js'
+  // behavior of truncating the response at the first invalid hex
+  // pair as opposed to just marking that an error happened and
+  // trying to continue with the decode.
+  if (text.size() % 2 != 0) {
+    text = text.slice(0, text.size() - 1);
+  }
+  kj::Vector vec = kj::Vector<kj::byte>(text.size() / 2);
+
+  for (auto i = 0; i < text.size(); i += 2) {
+    byte b = 0;
+    KJ_IF_MAYBE(d1, tryFromHexDigit(text[i])) {
+      b = *d1 << 4;
+    } else {
+      break;
+    }
+    KJ_IF_MAYBE(d2, tryFromHexDigit(text[i+1])) {
+      b |= *d2;
+    } else {
+      break;
+    }
+    vec.add(b);
+  }
+
+  return vec.releaseAsArray();
+}
+}  // namespace
+
+uint32_t BufferUtil::byteLength(jsg::Lock& js, v8::Local<v8::String> str) {
+  // Gets the UTF8 byte length for the given input. We use v8::String
+  // directly in order to avoid any copying and we just use the V8 API
+  // to determine the UTF8 length.
+  return str->Utf8Length(js.v8Isolate);
+}
+
+int BufferUtil::compare(
+    jsg::Lock& js,
```

Comment:
do we actually need this lock as parameter in methods like these, or did you just add it everywhere as a convenience?

Reply:
Convenience/Convention